File: kirun-java/src/main/java/com/fincity/nocode/kirun/engine/util/date/DateCompareUtil.java
package com.fincity.nocode.kirun.engine.util.date;

import java.util.Calendar;
import java.util.TimeZone;

import com.google.gson.JsonArray;

public class DateCompareUtil {

    private DateCompareUtil() {

    }

    public static final String YEAR = "years";

    public static final String MONTH = "months";

    public static final String DAY = "days";

    public static final String HOUR = "hours";

    public static final String MINUTE = "minutes";

    public static final String SECOND = "seconds";

    public static final String MILLIS = "millis";

    public static final String BEFORE = "before";

    public static final String AFTER = "after";

    public static boolean inBetween(String firstDate, String secondDate, String thirdDate, JsonArray fields) {

        return compare(firstDate, thirdDate, BEFORE, fields) && compare(secondDate, thirdDate, AFTER, fields);
    }

    public static boolean compare(String firstDate, String secondDate, String operationName, JsonArray fields) {

        Calendar firstCal = Calendar.getInstance(TimeZone.getTimeZone("UTC"));

        firstCal.setTimeInMillis(GetTimeInMillisUtil.getEpochTime(firstDate));

        Calendar secondCal = Calendar.getInstance(TimeZone.getTimeZone("UTC"));

        secondCal.setTimeInMillis(GetTimeInMillisUtil.getEpochTime(secondDate));

        boolean equal = true;

        for (int i = 0; i < fields.size(); i++) {

            equal = equal && compare(firstCal, secondCal, operationName, fields.get(i).getAsString());
        }

        return equal;

    }

    private static boolean compare(Calendar firstDate, Calendar secondDate, String operationName, String fieldName) {

        switch (fieldName) {

            case YEAR:
                return checkFields(firstDate, secondDate, operationName, Calendar.YEAR);

            case MONTH:
                return checkFields(firstDate, secondDate, operationName, Calendar.YEAR, Calendar.MONTH);

            case DAY:
                return checkFields(firstDate, secondDate, operationName, Calendar.YEAR, Calendar.MONTH,
                        Calendar.DAY_OF_MONTH);

            case HOUR:
                return checkFields(firstDate, secondDate, operationName, Calendar.YEAR, Calendar.MONTH,
                        Calendar.DAY_OF_MONTH,
                        Calendar.HOUR_OF_DAY);

            case MINUTE:
                return checkFields(firstDate, secondDate, operationName, Calendar.YEAR, Calendar.MONTH,
                        Calendar.DAY_OF_MONTH,
                        Calendar.HOUR_OF_DAY, Calendar.MINUTE);

            case SECOND:
                return checkFields(firstDate, secondDate, operationName, Calendar.YEAR, Calendar.MONTH,
                        Calendar.DAY_OF_MONTH,
                        Calendar.HOUR_OF_DAY, Calendar.MINUTE, Calendar.SECOND);

            case MILLIS:
                return checkFields(firstDate, secondDate, operationName, Calendar.YEAR, Calendar.MONTH,
                        Calendar.DAY_OF_MONTH,
                        Calendar.HOUR_OF_DAY, Calendar.MINUTE, Calendar.SECOND, Calendar.MILLISECOND);

            default:
                return false;
        }

    }

    private static boolean checkFields(Calendar firstDate, Calendar secondDate, String operationName, int... fields) {

        int fieldsLength = fields.length;

        if (fieldsLength == 0)
            return false;

        int equalLength = 0;
        int firstDateTotal = 0;
        int secondDateTotal = 0;

        for (int i = 0; i < fieldsLength; i++) {

            firstDateTotal += firstDate.get(fields[i]) * (i + 1);

            secondDateTotal += secondDate.get(fields[i]) * (i + 1);

            if ((i < fieldsLength - 1) && checkEquality(firstDateTotal, secondDateTotal, operationName)
                    || (i == fieldsLength - 1)
                            && checkEqualityForLastField(firstDate, secondDate, operationName, fields[i]))

                equalLength++;

            else
                break;

        }

        return equalLength == fieldsLength;
    }

    private static boolean checkEquality(int firstDateTotal, int secondDateTotal, String operation) {

        if (operation.equals("same"))
            return firstDateTotal == secondDateTotal;

        else if (operation.equals(BEFORE))
            return firstDateTotal <= secondDateTotal;

        else if (operation.equals(AFTER))
            return firstDateTotal >= secondDateTotal;

        return false;

    }

    private static boolean checkEqualityForLastField(Calendar firstDate, Calendar secondDate, String operation,
            int field) {

        if (operation.equals("same"))
            return firstDate.get(field) == secondDate.get(field);

        else if (operation.equals(BEFORE))
            return firstDate.get(field) < secondDate.get(field);

        else if (operation.equals(AFTER))
            return firstDate.get(field) > secondDate.get(field);

        return false;

    }

}

File: kirun-js/src/engine/util/date/DateCompareUtil.ts
import { KIRuntimeException } from '../../exception/KIRuntimeException';

const iso8601Pattern =
    /^([+-]?\d{6}|\d{4})-(0[1-9]|1[0-2])-(0[1-9]|[1-2]\d|3[0-1])T([0-1]\d|2[0-3]):([0-5]\d):([0-5]\d)(\.\d+)?(Z|([+-]\d{2}:\d{2}))?$/;

export class DateCompareUtil {
    public static readonly YEAR: string = 'year';

    public static readonly MONTH: string = 'month';

    public static readonly DAY: string = 'day';

    public static readonly HOUR: string = 'hour';

    public static readonly MINUTE: string = 'minute';

    public static readonly SECOND: string = 'second';

    private LOOPDURATION = 0;

    public inBetween(firstDate: string, secondDate: string, thirdDate: string, fields: []) {
        return (
            this.compareFields(firstDate, thirdDate, 'before', fields) &&
            this.compareFields(secondDate, thirdDate, 'after', fields)
        );
    }

    public compareFields(
        firstDate: string,
        secondDate: string,
        operationName: string,
        fields: [],
    ): boolean {
        let equal = true;

        let firstDateTime;
        let secondDateTime;

        for (let i = 0; i < fields.length; i++) {
            let currentAnswer = false;
            this.loopDuration(fields[i]);
            firstDateTime = this.getTimeInMilliSeconds(firstDate);
            secondDateTime = this.getTimeInMilliSeconds(secondDate);
            if (firstDateTime && secondDateTime) {
                if (operationName == 'same') {
                    currentAnswer = firstDateTime == secondDateTime;
                }
                if (operationName == 'after') {
                    currentAnswer = firstDateTime >= secondDateTime;
                }
                if (operationName == 'before') {
                    currentAnswer = firstDateTime <= secondDateTime;
                }
            }
            equal = equal && currentAnswer;
        }

        return equal;
    }

    private getTimeInMilliSeconds(date: string): number | null {
        const match = date.match(iso8601Pattern);
        let arr = [1970, 0, 0, 0, 0, 0];
        if (match) {
            for (let i = 1; i < this.LOOPDURATION + 1; i++) {
                arr[i - 1] = parseInt(match[i]);
            }
            const timeInMilliSeconds = new Date(arr[0], arr[1], arr[2], arr[3], arr[4], arr[5]);
            return timeInMilliSeconds.getTime();
        }
        return null;
    }

    private loopDuration(field: string) {
        if (field == DateCompareUtil.YEAR) {
            this.LOOPDURATION = 1;
        }
        if (field == DateCompareUtil.MONTH) {
            this.LOOPDURATION = 2;
        }
        if (field == DateCompareUtil.DAY) {
            this.LOOPDURATION = 3;
        }
        if (field == DateCompareUtil.HOUR) {
            this.LOOPDURATION = 4;
        }
        if (field == DateCompareUtil.MINUTE) {
            this.LOOPDURATION = 5;
        }
        if (field == DateCompareUtil.SECOND) {
            this.LOOPDURATION = 6;
        }
    }
}
